src/main/java/com/atlassian/db/replica/internal/LazyReference.java

```diff
@@ -1,31 +1,36 @@
 package com.atlassian.db.replica.internal;
 
 
-import java.util.concurrent.atomic.AtomicReference;
+import com.atlassian.db.replica.internal.util.ThreadSafe;
+
 import java.util.function.Supplier;
 
+@ThreadSafe
 public abstract class LazyReference<T> implements Supplier<T> {
-    private final AtomicReference<T> reference = new AtomicReference<>();
+    private T reference;
+    private final Object lock = new Object();
 
     protected abstract T create() throws Exception;
 
     public boolean isInitialized() {
-        return reference.get() != null;
+        return reference != null;
     }
 
     @Override
     public T get() {
-        if (!isInitialized()) {
-            try {
-                reference.compareAndSet(null, create());
```

Comment:
This CAS should have worked earlier too. If it returns `false`, then another thread was able to `create`, so we're good.
I ran `git checkout upstream/master -- src/main` and the test is still green.

- [ ] make sure the change is needed

Reply:
It's strange. The CAS should not work as create call is not protected by CAS. Also, the test was red for me, and I call it multiple times. The test also failed on CI (see the screenshot).
1. CAS will help us not to set the reference twice.
2. CAS will not protect from allocating a connection.

Reply:
You're right. Maybe I was lucky

Reply:
It's hard to write a concurrency test to reproduce a race condition that is not flaky and doesn't run forever 😢 .

Reply:
No worries 🙂 
It's still great that it reproduced a few times.